### DISCUSSION TITLE
Freestanding Val

### DISCUSSION THREAD

**Discussion by charlestolley** [hylo-lang/hylo#740] [Language design]
In order to make Val competetive with C and C++ in the long term, I think it's important to support bare-metal, embedded, and systems-level development. I think that the largest reason that C has remained such a dominant language for so many decades is that it is the easiest language to compile for these hostile runtime environments. Rust seems to have (somewhat uniquely) figured out this secret, and I think it's a major driver of its success. I love the commitment to building a high-quality standard library, and Swift's is a great one to model, but I'm concerned that an oversized standard library will limit the reach of Val and keep it perpetually dependent on a C and C++ runtime environment. I've heard old interviews with Chris Lattner in which he mentioned that Swift could be used for writing low-level device drivers, but it doesn't seem to have ever caught on, and I'm not sure his claim is still true today (I once tried to compile Swift for a bare-metal system and was unsuccessful).

The C and C++ standards demarcate exactly which features must be supported for freestanding versus hosted environments, I think Val should consider doing something similar.

**Comment by dabrahams** [comment]
We're planning to try to do something like [Rust's killer feature](https://docs.rust-embedded.org/book/intro/no-std.html).  In our case the core library is lower-level than the standard library, and  just big enough to supply types like `Int` that would otherwise be built-in in a language like C or C++.

**Comment by khaledh** [comment]
Apologies for resurrecting this thread from the dead. I'm interested in low-level systems programming, especially hobby kernel/OS development. I tried C, Zig, and Nim, and I'm exploring other languages as well. I'm interested in understanding where Hylo is w.r.t. this domain.

**Reply by dabrahams** [reply]
Nothing has changed; it's still in the plan.  We have a compiler flag that turns off the parts of the standard library that depend on a full-featured OS.  There may well be more that could/should be done, but to know what that is, we need to hear from practitioners in the domain about their needs.

**Reply by khaledh** [reply]
Thanks @dabrahams. I'm still getting my feet wet with Hylo. Once I have a grasp of the compiling/linking process, and where it depends on OS services, I'll report back my findings.

**Reply by mrael2** [reply]
It would be nice to use Hylo installed on my M1 Macbook Air to cross-compile to RISC-V for use on, say, the StarFive VisionFive 2 single-board computer. Would it be possible to have a flag which produces assembly language output in addition to the binary to see what RISC-V instructions were chosen by the Hylo compiler?

**Reply by dabrahams** [reply]
If you implement it, sure!  Hylo generates LLVM, and we haven't done anything explicit yet to support cross-compilation.
